### ISSUE TITLE
Questions, and wondering how best to collaborate on a single PhysX WASM port

### ISSUE THREAD

**Issue by prestomation** [luisfonsivevo/physx.js#2]
Hi Luis,

Awesome to see another PhysX WASM port. Mine is here, which I think you've seen: https://github.com/prestomation/PhysX/tree/emscripten_wip

I had a look at your code. Lots of things are pretty similar. 

I see you're not using `allow_raw_pointer()`. I spent quite some time early on trying to get this to work and was unsuccessful. What's your secret? :)

I also noticed you are calling `release()` in the destructor overrides. Emscripten can't doesn't/can't call destructors right?  This would mean that library consumers would call `.delete()`(an emscripten function) which would call release and then delete the memory. For my consumers, they call `release()` just like they would in PhysX, and PhysX takes care of deleting objects, or at least that's how I thought it was working. I'm wondering if you had any more thoughts here? C++ is not my forte and this is my first emscripten project so I could be totally off.

All that said, I'm thinking we should chat about how to collaborate on a single PhysX project. One problem I saw with ammojs was the community is somewhat fragmented, the 'de facto' build is using an old fork of bullet , and various people have forked it to upgrade and it's all kind of a mess. This is why I chose to fork PhysX itself and build emscripten support in as close as a native platform as possible, with the hopes that Nvida takes this upstream. If they don't, then I would build some sort of community around the repo and it would be easier to merge in changes from PhysX. It seems less good for the ecosystem as large to fragment this, I just wish we had learned about each others projects sooner! What are your thoughts?

**Comment by luisfonsivevo** [comment]
I think it would be awesome if we could merge projects somehow. Fragmentation never helped anyone. 

- I like how you were able to add your own emscripten platform to the existing PhysX build system. That will definitely help get Nvidia's interest in supporting us upstream. I wish I could've done that but I went in knowing nothing about cmake.
- Having to write `allow_raw_pointer()` for everything was annoying me so I removed the error message from the emscripten headers and shipped my own. This is what I changed in wire.h:
```
template<typename T>	
struct TypeID<T*> {	
    //BEFORE
    static_assert(!std::is_pointer<T*>::value, "Implicitly binding raw pointers is illegal.  Specify allow_raw_pointer<arg<?>>");
    //AFTER
    return LightTypeID<T*>::get();
};
```
- For some reason embind duplicates objects and object references if a function or property getter returns them. Did you come across this bug in embind? https://github.com/emscripten-core/emscripten/issues/9930 I was having horrible memory leaks trying to do the most basic tasks and I wonder how you got around it. I had to modify the emscripten headers again (in addition to the`allow_raw_pointer()` modification) to work around it, using the hotfix I mentioned in the replies.
- Emscipten's `delete()` function in js code calls `raw_destructor<PxWhatever>()` in C++ which defaults to calling that class's destructor. Whenever you call `new PxWhatever()` in jsland you eventually have to `delete()` that object when you're done with it. The problem that we both found is that embind defaults to calling PhysX's private destructors, so we couldn't compile without overriding `raw_destructor`. You chose to have `raw_destructor()` do nothing and require the library consumer to call `release()` instead, while I overrode `raw_destructor()` with `release()` itself because I was too lazy to remember which classes need to be deleted vs released. The only difference is just personal preference: sticking to the API vs. convenience. I think the proper way to do this would be to throw a console error whenever `delete()` is called on an object with a private destructor, and expose `release()` through embind like you did. And yes you are correct about how `release` is supposed to work(). PhysX automatically deletes the object when the reference count hits 0 and you've called `release()`.
- The biggest problem I've been having is with raw C++ arrays and function callbacks. I saw you got around the array problem using embind's `register_vector` so I guess I just didn't read the docs thoroughly enough. But we both have kind of hacky workarounds for function callbacks (for example PxDefaultSimulationFilterShader). I was hoping there was a better way to do it.
- I think the best thing to do from here is to agree on how to handle memory management and any other small differences, then I can start merging my embind.cpp into your PxWebBindings.cpp and hopefully Nvidia accepts your PR.

Thanks for checking out my project. I hope a lot of people can benefit from a collaboration. B-)

**Comment by prestomation** [comment]
Thanks for all the details.

Now I understand why you had embind code in repo. I'm not much of a fan of copy-and-modifying code like this, as you never know what will break with new builds of emscripten as we'd have to manage updates to these files and ideally track a particular version/etc.. Seems opening a PR to emscripten so we can set a compile-time define to do this is a good option though.


Our port uses value objects for transform/vec/quat. This means they are just data. In our use case in Sumerian, this reduced code and copies as we were just going to translate these to the equivalent types in our code anyway, and removes the need to call delete. My hypothesis was that consumers of this would be a similar situation and be using their own versions of these concepts in whatever framework they are using and so use those vec libraries if they needed the math functions. This would mitigate the issue you describe when it comes to these types. 

For memory management, users still need to be able to call release() on many of these objects separately from deleting them. This allows consumers to share geos/materials/etc and not worry about when to delete them, and just worry about releasing references when removing a particular usage. This was why I decided to stick to PhysX's API as much as possible.

`PxDefaultSimulationFilterShader` was something I scratched my head over for a while and couldn't come up with a better answer. Something someone can spend some more time digging into I suppose. There are number of similarly ugly things in our initialization that shouldn't be there, I'll clean that up.

**Comment by luisfonsivevo** [comment]
I agree that I should delete the modified headers at some point but right now I'm dependent on them. I'm going to submit an emscripten PR tomorrow to add a `#EMSCRIPTEN_ALLOW_ALL_RAW_POINTERS` define so I can ease away from needing the embind headers.

I tried turning `Vec3` and `Quat` into value objects (I still have `PxTransform` as a class for now) but I'm having a weird problem. I try to access `PxTransform.p` (looks like you renamed yours to `PxTransform.translation`) so I can update the 3D model position every frame, but as soon as PhysX detects a collision between some actors, I can no longer use the `PxTransform.p` getter without a memory access error:
```
Uncaught RuntimeError: memory access out of bounds
    at wasm-function[68]:0x5eab
    at wasm-function[75]:0x65fc
    at wasm-function[2356]:0x147e63
    at Module.dynCall_vi (http://furrygamingpc/physx.js/build/physx.js:2530:39)
    at dynCall_vi_2865 (eval at makeDynCaller (http://furrygamingpc/physx.js/build/physx.js:1361:24), <anonymous>:4:12)
    at Object.fromWireType (http://furrygamingpc/physx.js/build/physx.js:746:25)
    at Transform.get [as p] (http://furrygamingpc/physx.js/build/physx.js:1688:68)
    at updateModelTransform (http://furrygamingpc/physx.js/demo/simplegl/:256:20)
    at draw (http://furrygamingpc/physx.js/demo/simplegl/:205:3)
```
I don't know if that's something you came across but I gave up for the night.

In other news, I added `release()` methods to all the classes with private destructors and made `delete()` throw an error message that tells the library consumer to use `release()` instead. There shouldn't be any need to ever call `delete()` on these classes because PhysX takes care of it for you once you `release()`.

Next I want to remove my helper classes `JSMeshDesc` and `JSArray` and clean up a few `select_overloaded`ed functions, then my bindings should be pretty much conformant to the C++ API. I notice you have a lot of helper functions at the top of your bindings. Are you planning on keeping those in?

**Comment by prestomation** [comment]
I have not seen this memory access out of bounds error.

I do not want to keep these helpers if we can make what they are trying to do as much like the C++ API as possible. These were added by others to help our particular usecase, but I do want to make them more like the C++ API.  Same with `PxTransform.translation`, wish we hadn't had done that.
